libs/pilight/core/common.c

```diff
@@ -1019,3 +1024,161 @@ int file_get_contents(char *file, char **content) {
+#endif
+
+// Do NOT write the next as range A-Za-z0-9 because the string may be used without RE.
+#define AtoZ_atoz_0to9 "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+#define white_spaces " \t"
+
+// Next email_local_part_valid_chars is defined in a way that it can be used without RE
+// and within [] of a posix extended regular expression. Therefore the dash comes first
+// because at later position it would mean range (and it can't be escaped because
+// a \ inside [] is literal - this is different from perl RE).
+#define email_local_part_valid_chars "-" AtoZ_atoz_0to9 ".!#$%&'*+\\/=?^_`{|}~"
+
+#if use_regexp
+// The RE to check email addr is compiled only once and used commonly.
+
+static regex_t check_email_common_RE = { 0 };
+static const regex_t * volatile check_email_compiled_RE = NULL;
+
+static void check_email_free_RE() {
+	if(check_email_compiled_RE != NULL) {	// protect against unexpected calls.
+		regfree(&check_email_common_RE);
+		check_email_compiled_RE = NULL;
+	}
+}
+#endif
+
+/**
+ * Check if an email address is valid.
+ * @param char* addr The email address to check. Leading+trailing white spaces are accepted.
+ * @param bool allow_lists If true, addr is allowed to be a comma separated list of email addresses.
+ * @param bool check_domain_can_mail If true, check if domain exists in DNS system and accepts email.
+ * @return int Values < 0 indicate error. Values >= 0 indicate success.
+ * Where -1: syntax error, -2: DNS check for a domain failed, -8: severe internal error.
+ */
+int check_email_addr(const char *addr, int allow_lists, int check_domain_can_mail) {
+	if(addr == NULL) {
+		return -1;
+	}
+#if use_regexp
+	const regex_t *RE = NULL;
+	int reti = 0;
+
+	static const char email_pattern[] =
+		"^[" email_local_part_valid_chars "]+"
+		"@("
+			   "([A-Za-z0-9][-A-Za-z0-9]*[A-Za-z0-9])"
+			"(\\.[A-Za-z0-9][-A-Za-z0-9]*[A-Za-z0-9])+"
+		"|"
+			"[0-9]{1,3}(\\.[0-9]{1,3}){1,3}"	// IPv4 addr
+		")$";
+
+	if((RE = check_email_compiled_RE) == NULL) { // first check without lock.
+		// RE not compiled yet.
+		atomiclock();
+		if((RE = check_email_compiled_RE) == NULL) { // locked now, need to check again.
+			reti = regcomp(&check_email_common_RE, email_pattern, REG_EXTENDED);
+			if(reti == REG_NOERROR) {
+				check_email_compiled_RE = RE = &check_email_common_RE;
+				atexit(check_email_free_RE);
+			} else {
+				char errmsg[regerror(reti, &check_email_common_RE, NULL, 0)];
+				regerror(reti, &check_email_common_RE, errmsg, sizeof(errmsg));
+				regfree(&check_email_common_RE);
+				atomicunlock();	// do not logprintf while locked.
+				logprintf(LOG_ERR, "could not compile regex pattern '%s' to check email addresses: %s",
+						email_pattern, errmsg);
+				return -1;
+			}
+		}
+		atomicunlock();
+	}
+#else
+	static const char local_valid_chars[] = email_local_part_valid_chars;
+	static const char domain_valid_chars[] = AtoZ_atoz_0to9 ".-"; // . and - are not allowed everywhere.
```

Comment:
yes, local_valid_chars could be eleminated, its used only once. I indeed already thought about not to define this buffer. Finally I introduced the name because it does not cost anything and has the benefit that there is a name for the charsets that can be examined when debugging.